### DISCUSSION TITLE
Reworking builtin/bare

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [vlang/v#11841] [Ideas]
Currently, `v -freestanding` is broken, because `builtin/bare` hasn't received much attention in a while.

So, this is a proposal to rework it. I am currently looking into this.

The questions that `builtin/bare` should answer:
- Should it focus on embedded devices (like the Raspberry Pi Micro and other microcontrollers) or should it be ment for regular OS uses? Alternating between the two is difficult because:
  * The current `map` implementation uses `strconv` (which is huge) and wyhash
  * `chan` is most often not going to be supported on those targets at all - either due to a lack of threading or specific threading models that might not be implemented. But the former is most often the case with the latter being specific to each platform.
- Should it support **all** convenience methods and features present in the standard builtins? For instance, `'1'.int()` calls strconv methods and thus becomes quite big and produces overhead. Is this really neccessary?

For the first, I think it should look to being as minimal as possible so that embedded devices can take advantage of V's features but it should not rely on all the convenience methods either.

This is what I think `builtin/bare` should do:

- `string`
  * Wrap `char*` for the most part - which is what it already does.
  * Strip all convenience methods. (`.bool()`, `.int()`, `.f32()`, ...)
- `map`
  * Use a much more basic approach.
  * This approach will, obviously, be slower and less optimized.
  * Support the basic iteration methods (`.map()`, `.filter()`, ...)
- `chan`
  * Implement it on a per-platform basis and otherwise stub it.
  * One could use `chan_$os.c.v` to implement the native threading - if available - to support this type.
- `float`
  * I am no expert on floats. But do we really need four stringification methods?
- `utf8`
  * Nice to have - but I don't think it's really neccessary. I'd wager to not include it here.
- `builtin`
  * Most of the methods here have to do with debugging and printing debug information, if compiled as such.
  * Each platform quite likely has a different way of outputting debug information, getting stack traces and the likes. Thus, this should possibly be renamed into `debug` instead and implemented for each target differently.
  * There are a lot of structures related to reflection. On a small device, that might actually be overhead. I am unsure about that since it relates to asserting. So for now, I am leaning towards reworking asserts into something more simple but stripping some of the reflection.

The biggest drivers in binary size are `strconv` and `map` - those together also require wyhash, which makes just these three alone impressively huge. Easily too huge for what a small device might be able to handle.

I thought I'd post my thoughts here in order to see what other people think and to find a direction for this. Since I am trying to bring V to other platforms and targets, this is something I am particularily interested in. Plus, the Small Device C Compiler (SDCC) does not support everything in terms of standard C. It *does* support C99 fine, but it can not pass struct fields as arguments and also currently can not handle compound statements (`func(&arg)`). Since SDCC is ment to compile to various smaller devices, I thought it to be a great way to try and see if V could run on the devices it targets.

Hope to hear back from you guys. :)

**Comment by wangyi-fudan** [comment]
BTW: wyhash has been improved significantly. try the latest version.

**Comment by JalonSolov** [comment]
@hungrybluedev ^^

**Comment by hungrybluedev** [comment]
I think this is a step in the right direction. I believe @ka-weihe and @wangyi-fudan (creator of wyhash) should discuss about reworking the hashing implementation of maps.

For `strconv` and other conversions, I think @penguindark is the right person to ask.

We definitely do need a Rune type or something equivalent. I would (personally) prefer a utf8 implementation rather than a ut16 version.

I agree with the fact that channels are difficult to implement for embedded devices.  Focus should be on simple, single threaded core libraries and layering channelling utilities on top of it. We definitely will not neglect users of regular OSes but targeting embedded devices will probably be an easier target. I must admit @spytheman definitely knows more than I do about this aspect. I would also like to bring @spaceface777 into the conversation.

About the removal of convenience methods, I'm not sure about how to figure out what's essential and what's not. Careful planning must be done in this regard.

I have not addressed all the points in the discussion, feel free to @ me or ask others about specific topics.

**Comment by penguindark** [comment]
About `f32/f64` to string and viceversa, if anyone have a better and precise and faster algos, I'l be happy to implement it.
At the present time V has the fastest conversions f32/f64 to and from string available on the leterature.

They are designed to avoid to use stdlib and overcome some bugs in the C stdlib and to allow to use V even on bare metal without need to link C libraries at all.
(not the case at the present time ;) )

For sure they need disk space but if (as we expect ;) ) the unused data and function will not be linked in future the size will be for sure reduced.

If you are curious about sizes, please, have a look at the C f32/f64 to string conversion ;)

**Comment by ka-weihe** [comment]
> reworking the hashing implementation of maps

why? Wyhash is perfect.

> BTW: wyhash has been improved significantly. try the latest version.

I actually planned to update to the newest version today :slightly_smiling_face: 

> The biggest drivers in binary size are strconv and map - those together also require wyhash, which makes just these three alone impressively huge

Wyhash is tiny. The map is also tiny. With a couple of updates to skip-unused the produced code should be quite small.

I'm not buying that the map is "huge".

**Comment by JalonSolov** [comment]
I agree with the need to have a 'bare' implementation, but I do think we need things like UTF-8 support, as well as channels, etc.

If you strip too much of V, why use V at all?

I think it will be more important to get `-skip-unused` finished and made the default, so that these extra features will be stripped when not used, but will still be available.

**Comment by IngwiePhoenix** [comment]
I am gone for two days and my inbox went and bulked up some. Huh. o.o

Thank you guys for pointing out wyhash's improvements. I also re-read the map implementation just to make sure - and, i have to confess that I might have overreacted a little. What I percieved as "big" was the sheer amount of code for it. That was a misunderstanding on my part. Re-reading it again actually clears it up quite some. I haven't yet read the entirity of wyhash to make myself a better picture of it, but I'll take your word there. The CRC32 implementations of hashmaps in a single C header that I have found were small, but only worked with string keys - adopting them to other datatypes would probably make the size increase again. So; tl;dr is that I was overreacting about the size of `map`. Sorry! ^^;

`strconv` on the other hand...is impressively big, but for good reasons too. As for `-skip-unused`: While doing my GBA proof of concept of running V on the GBA, `strconv` takes an enormous amount of memory. Yes - most microcontrollers and embedded devices these days probably have more than the GBA in terms of work memory but this shouldn't be underestimated either. The GBA might be an edge case in the bigger picture, but if it can run V with the current builtin implementation, then a slimmed down builtin/bare should help there a whole lot.

So from reading all the answers, it seems that the first thing to focus on would be to built a copy of the regular builtin and remove `chan` and dependencies to `strconv`. Would make a nice start for the beginning. Anything else that can be stripped afterwards should be looked at in individual cases.

@hungrybluedev 

> About the removal of convenience methods, I'm not sure about how to figure out what's essential and what's not. Careful planning must be done in this regard.

I'd go by this: Does *this* method matter to the developer, or the user? For instance, printing pretty numbers is most relevant when outputting data to the user. But, a microcontroller, like the Raspberry Pi Micro, won't have a "readable" output (as in, it's less likely a user would read the raw stdout/-err output of it). So methods like `'1.2'.f32()` are less likely to be needed.

A compromise would possibly be to work with `-d`s. Have a `string_d_use_stringify.v` (had to make that name up) that *does* use `strconv` to offer those methods - and return an empty string (or use `#error`?) when the methods are called but the feature is not enabled.

In general, since baremetal targets might have more or fewer resources, giving users the option to opt in - or out - of some features might be handy. A GBA might not need the entirity of f64 - it just is not that precise - but something like a Raspberry Pi Micro doing meassurements totally would want this. I think there is something to be said about offering a little more flexibility for the baremetal target. Basically, `-d` shouldn't be too needed for "normal V development" but possibly be leveraged for baremetal to customize the output. Just a thought, really :)

**Comment by hungrybluedev** [comment]
@IngwiePhoenix it seems we'd benefit from a modular system where users can choose what parts of the VLib they want to use for the target.

Before that, there should be a core subset identified, that we expect to work on all targets. The other modules should be easy to plug in and use.

I'd recommend writing an outline/proposal for the baremetal core and posting it here.

**Comment by IngwiePhoenix** [comment]
Strings, arrays, maps - those are the core features of V that are not generated by the compiler, so those are a nobrainer in terms of needing to be there.

Int and float - could possibly be simplified for the core?

chan - as discussed above, this should be optional for the implementer to add since this is quite device-specific.

debug - Also device specific (some have an actual stdout/stderr - but some others print to a serial connection). So this should be optional and implemented per device.

It's about 6am; I'll write a better, more detailed list later. But this should do as a short note. If you have any suggestions there, feel free to post them! I'll take them into account :)

**Comment by penguindark** [comment]
@IngwiePhoenix about `strconv` if you need a simple string to float conversion, there is already a quick and short function:
https://github.com/vlang/v/blob/master/vlib/strconv/atofq.v
you can use it without include all the strconv ;)

**Comment by JalonSolov** [comment]
Most of vlib is already "pick & choose"... if you import it and use it, you've "picked" it.  I don't see any need to introduce new options to control the same thing.

Same thing with channels... either use them or don't.

Things like those will be handled better as `-skip-unused` is improved.

"Heavy" routines like strconv... I can see where we might like a bit more control over whether or not those got used, but it seems like extra `-d` options is a clumsy approach to this.  There are a lot of smart folks working on V... perhaps a wider audience, to get more input?

**Comment by dumblob** [comment]
One more note about channels and/or alike. I think this is pretty useful for embedded systems even of the size slightly "bigger" than GBA. There is a message passing primitive for extremely memory&disk space limited uCs - namely special zero-copy channel-like [message passing primitive](https://github.com/betrusted-io/xous-core/issues/4#issuecomment-721018776 ) with borrow/... (as in Rust) semantics: [overview](https://github.com/betrusted-io/xous-core/blob/master/services/README.md ) and [technical information](https://github.com/betrusted-io/xous-core/blob/hardware/docs/processes.md ) (and [details about "push" events](https://github.com/betrusted-io/xous-core/wiki/Push-Events-and-Listeners )).

So IMHO it's not that clear channels/... should be somehow skipped by default for highly constrained embedded devices. I'd even argue the opposite :wink:.

**Comment by crthpl** [comment]
Is this fixed now that `-freestanding` works again? #9624
